idaes/generic_models/properties/core/eos/virial.py

```diff
@@ -0,0 +1,797 @@
+                                       'for i-j molecular pair'))
+
+        def rule_Pc_ij(m, i, j):
+            return (m.Zc_ij[i, j] * m.Tc_ij[i, j] / m.Vc_ij[i, j] *
+                    Virial.gas_constant(m))
+        b.add_component('Pc_ij',
+                        Expression(b.component_list,
+                                   b.component_list,
+                                   rule=rule_Pc_ij,
+                                   doc='Critical pressure combining rule '
+                                       'for i-j molecular pair'))
+
+        def rule_B_ij(m, i, j):
+            return (m.Tc_ij[i, j] / m.Pc_ij[i, j] * Virial.gas_constant(m) *
+                    (f_BO(m.temperature / m.Tc_ij[i, j]) +
+                     f_B1(m.temperature / m.Tc_ij[i, j]) * m.omega_ij[i, j]))
+        b.add_component('B_ij',
+                        Expression(b.component_list,
+                                   b.component_list,
+                                   rule=rule_B_ij,
+                                   doc='Second virial coefficient for '
+                                       'i-j molecular pair'))
+
+    @staticmethod
+    def calculate_scaling_factors(b, pobj):
+        pass
+
+    @staticmethod
+    def build_parameters(b):
+        param_block = b.parent_block()
+        # virial eos supports only vapor phase
+        if not b.is_vapor_phase():
+            raise PropertyNotSupportedError(f"{b.parent_block().name} "
+                                            "received unrecognized phase name"
+                                            f"{b}. Virial equation of"
+                                            "state supports only Vap phase.")
+
+        # get user specified eos options and save on parameter block
+        try:
+            use_pseudocritical_rules = b.config.equation_of_state_options[
+                "use_pseudocritical_rules"]
+            b._use_pseudocritical_rules = use_pseudocritical_rules
+            setattr(b.parent_block(), "virial_eos_options",
+                    deepcopy(VirialConfig))
+            ConfigBlock = getattr(b.parent_block(), "virial_eos_options")
+            ConfigBlock.set_value(b.config.equation_of_state_options)
+        except TypeError:
+            use_pseudocritical_rules = True
+            b._use_pseudocritical_rules = use_pseudocritical_rules
+            setattr(b.parent_block(), "virial_eos_options",
+                    deepcopy(VirialConfig))
+            ConfigBlock = getattr(b.parent_block(), "virial_eos_options")
+            ConfigBlock.set_value(b.config.equation_of_state_options)
+
+        # Get critical volume of components
+        units = b.parent_block().get_metadata().derived_units
+        for comp in b.parent_block().component_list:
+            cobj = b.parent_block().get_component(comp)
+            if not hasattr(cobj, 'volume_crit'):
+                setattr(cobj, "volume_crit",
```

Comment:
This is not the correct way to do this. `volume_crit` should be created by the component if the data is available - i.e this should already exist, or the user forgot to provide the data. In the second case, we should be raising an Exception, rather than trying to create the Var (as the data obviously doesn't exist). If the code does not yet exist to automatically build `volume_crit`, then that should be done elsewhere.